File: tensorflow/contrib/learn/python/learn/dataframe/transforms/densify.py
# Copyright 2016 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

"""Transforms Sparse to Dense Tensor."""

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

from tensorflow.contrib.learn.python.learn.dataframe import transform
from tensorflow.python.ops import sparse_ops


class Densify(transform.TensorFlowTransform):
  """Transforms Sparse to Dense Tensor."""

  def __init__(self,
               default_value):
    super(Densify, self).__init__()
    self._default_value = default_value

  @transform.parameter
  def default_value(self):
    return self._default_value

  @property
  def name(self):
    return "Densify"

  @property
  def input_valency(self):
    return 1

  @property
  def _output_names(self):
    return "output",

  def _apply_transform(self, input_tensors, **kwargs):
    """Applies the transformation to the `transform_input`.

    Args:
      input_tensors: a list of Tensors representing the input to
        the Transform.
      **kwargs: Additional keyword arguments, unused here.

    Returns:
        A namedtuple of Tensors representing the transformed output.
    """
    s = input_tensors[0]

     # pylint: disable=not-callable
    return self.return_type(sparse_ops.sparse_to_dense(
        s.indices, s.dense_shape, s.values, default_value=self.default_value))

File: tensorflow/lite/kernels/densify.cc
/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/
#include "tensorflow/lite/kernels/internal/reference/densify.h"

#include <stddef.h>

#include <cstdint>

#include "tensorflow/lite/core/c/common.h"
#include "tensorflow/lite/kernels/internal/tensor_ctypes.h"
#include "tensorflow/lite/kernels/kernel_util.h"

namespace tflite {
namespace ops {
namespace builtin {
namespace densify {

struct OpContext {
  OpContext(TfLiteContext* context, TfLiteNode* node) {
    input = GetInput(context, node, 0);
    output = GetOutput(context, node, 0);
  }
  const TfLiteTensor* input;
  TfLiteTensor* output;
};

struct OpData {
  bool dense_weights_initialized;
};

void* Init(TfLiteContext* context, const char* buffer, size_t length) {
  auto* op_data = new OpData();
  op_data->dense_weights_initialized = false;
  return op_data;
}

void Free(TfLiteContext* context, void* buffer) {
  delete reinterpret_cast<OpData*>(buffer);
}

TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
  TF_LITE_ENSURE_EQ(context, NumInputs(node), 1);
  TF_LITE_ENSURE_EQ(context, NumOutputs(node), 1);

  OpContext op_context(context, node);

  TF_LITE_ENSURE(context, op_context.input != nullptr);
  TF_LITE_ENSURE(context, op_context.input->type != kTfLiteString);
  TF_LITE_ENSURE(context, IsConstantTensor(op_context.input));
  TF_LITE_ENSURE(context, op_context.input->sparsity != nullptr);

  op_context.output->type = op_context.input->type;
  op_context.output->name = "Densify_output";
  op_context.output->allocation_type = kTfLiteArenaRwPersistent;

  return context->ResizeTensor(context, op_context.output,
                               TfLiteIntArrayCopy(op_context.input->dims));
}

TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
  OpData* op_data = reinterpret_cast<OpData*>(node->user_data);
  OpContext op_context(context, node);
  if (op_data->dense_weights_initialized) {
    return kTfLiteOk;
  }

  switch (op_context.input->type) {
    case kTfLiteFloat32:
      reference_ops::Densify(op_context.input->sparsity,
                             GetTensorShape(op_context.input),
                             GetTensorData<float>(op_context.input),
                             GetTensorShape(op_context.output),
                             GetTensorData<float>(op_context.output), context);
      break;
    case kTfLiteFloat16:
      reference_ops::Densify(
          op_context.input->sparsity, GetTensorShape(op_context.input),
          GetTensorData<Eigen::half>(op_context.input),
          GetTensorShape(op_context.output),
          GetTensorData<Eigen::half>(op_context.output), context);
      break;
    case kTfLiteInt8:
      reference_ops::Densify(op_context.input->sparsity,
                             GetTensorShape(op_context.input),
                             GetTensorData<int8_t>(op_context.input),
                             GetTensorShape(op_context.output),
                             GetTensorData<int8_t>(op_context.output), context);
      break;

    default:
      TF_LITE_KERNEL_LOG(context, "Type %d not supported.",
                         op_context.input->type);
      return kTfLiteError;
  }

  op_data->dense_weights_initialized = true;
  return kTfLiteOk;
}

}  // namespace densify

TfLiteRegistration* Register_DENSIFY() {
  static TfLiteRegistration r = {densify::Init, densify::Free, densify::Prepare,
                                 densify::Eval};
  return &r;
}

}  // namespace builtin
}  // namespace ops
}  // namespace tflite
